File: Problems/No.8_String_to_Integer_(atoi)/string_to_integer_(atoi).py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        # store the answer and the flag or positive, negative
        StI_ans = ""
        pos_neg = None
        
        # start compute
        for c in s:
            if (c == ' ' and pos_neg == None and StI_ans == ""):
                continue
            elif (c == '+' and pos_neg == None and StI_ans == ""):
                pos_neg = 0
            elif (c == '-' and pos_neg == None and StI_ans == ""):
                pos_neg = 1
            elif (c.isdigit()):
                StI_ans += c
            else:
                break
        
        # check if the final value is legal
        try:
            final_value = int(StI_ans)
            return max(-final_value, -2**31) if pos_neg else min(final_value, 2**31-1)
        except:
            return 0
            
                
        

File: Problems/No.8_String_to_Integer_(atoi)/string_to_integer_(atoi).cpp
class Solution {
public:
    int myAtoi(string s) {
        
        // ASCII code 48..."0", 57..."9"
        // INT_MAX ...  2147483647
        // INT_MIN ... -2147483648
        
        int c = 0;
        int StI_ans = 0;
        bool pos_neg = 0;
        
        // start compute
        // find until '-', '+', or digit appear
        while (c < s.length() and s[c] == ' ') c++;
        if (s[c] == '-' or s[c] == '+') {
            
            pos_neg = (s[c]=='-') ? 1 : 0;
            c++;
            
        }
        
        // compute the answer
        while (c < s.length() and 48 <= s[c] and s[c] <= 57) {
            
            int digit = s[c] - 48;
          
            // check whether it overflow or not
            if (!pos_neg and 
               ((StI_ans > 214748364) or 
                (StI_ans == 214748364) and (digit >= 7))) return INT_MAX;
            else if (pos_neg and
                    ((StI_ans > 214748364) or 
                     (StI_ans == 214748364) and (digit >= 8))) return INT_MIN;
            
            StI_ans = StI_ans*10 + digit;
            c++;
            
        }
        
        if (pos_neg) return -1*StI_ans;
        else return StI_ans;
        
    }
};
